### ISSUE TITLE
[Alerting V2] UI for PrometheusRules CRs

### ISSUE THREAD

**Issue by aiyengar2** [rancher/dashboard#811]
Need to create a UI for PrometheusRules CRs to make it easier for user to be able to create them.

**Description:** Generates alerting / recording rules based on rule groups.

### Example of PrometheusRules CR

Here is an example of a `PrometheusRules` CR that defines a single Alert Group (`example.rules`) with both an alerting rule (`Example Alert`) and a recording rule (`example_record`).

```yaml
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  name: prometheus-example-rules
  namespace: cattle-monitoring-system
  labels:
    app: rancher-monitoring
    release: rancher-monitoring
spec:
  groups:
  - name: example.rules
    rules:
    - alert: ExampleAlert
      expr: vector(1)
      annotations:
        message: Hello World!
      labels:
        severity: none
        job: myJob
    - record: example_record
      expr: vector(1)
      labels:
        job: myJob
```

Note: You can have >1 group in one PrometheusRules CR and each group can have >1 rules that are either alerting rules or recording rules (based on whether `alert` or `record` is used).

### User Experience

By default, when the user enters the page to create a PrometheusRule, they should see an empty Rule Group and a  `+` button that allows them to define one or more Rule Groups. There should also be a `Remove` button on each Rule Group to allow them to be removed and a user must have 1 or more Rule Groups defined for a PrometheusRule to be saved.

Each RuleGroup should follow the following spec:
Header / Label Name | k8s YAML | Description | Input | Default
---|---|---|---|---|
Group Name | `spec.groups[i].name` | `N/A` | `string` (required field) | `nil`
Override Group Interval | `spec.groups[i].interval` | How often rules in the group are evaluated. Defaults to global evaluation interval | `string` | nil
Rules | `spec.groups[i].rules` |  `N/A` | List of `Rules` |  `[]` with ability to add rules with `+` button and remove with `Remove` button

The `Rules` themselves take one of two forms. When clicking on the button to `Add Rule`, a user should be able to select which type of rule they want to create with the following help text available in some form:

Rule Type | Description |
---|---|
Create Alerting Rule | Defines an alert condition based on a PromQL query to send notifications to a configured AlertManager
Create Recording Rule | Defines a frequently needed or computationally intensive PromQL query that should be stored as a new set of time series

Based on which rule is added, the following fields should be exposed.
### AlertingRule

Help Text:
```
When firing alerts, the annotations and labels will be passed to the configured AlertManagers to allow them to construct the notification that will be sent to any configured Receivers.
```

Header / Label Name | k8s YAML | Description | Input | Default
---|---|---|---|---|
Alert Name | `*.alert` |  `N/A` | `string` (required) | `nil`
PromQL Expression | `*.expr` | A PromQL Expression used to check if this alert is firing | `string` (required) | `nil`
Wait To Fire Alert For | `*.for` | How long an alert should be firing for before a notification is sent out | String | `0s`
Severity | `*.labels.severity` | A Label that identifies the severity of this alert (warning, critical, none, etc.) | `string` | `none`
Summary | `*.annotations.summary` | A concise description of the state of the cluster once this alarm is triggered | `string` | `nil` 
Message | `*.annotations.message` | An explanation of what this alert means | `string` | `nil`
Description | `*.annotations.description` | A detailed description of what symptoms in the cluster triggered this alert | `string` | `nil`
Runbook URL | `*.annotations.runbook_url` | A URL that points to an external runbook used to resolve this alert | `string` | `nil`
Additional Labels | `*.labels` | Values used to identify an alert. These are automatically provided by the PromQL expression but can be overridden if provided here | `key`: `value` inputs of `strings` | See `Default Labels` below
Additional Annotations | `*.annotations` | Values used enrich the data provided by an alert when notified. It is possible to reference label values here via templates | `key`: `value` inputs of `strings` | See `Default Annotations` below

For Labels, a one-line `string` value should be shown. For annotations, it's possible that the user would submit a multi-line `string` value since it's expected to be used for longer additional information such as alert descriptions or runbook links.

If no entry is provided for the value, the key-value pair should be omitted.

### RecordingRule

Header / Label Name | k8s YAML | Description | Input | Default
---|---|---|---|---|
Time Series Name | `*.record` |  `N/A` | `string` (required) | `nil`
PromQL Expression | `*.expr` | `N/A` | `string` (required) | `nil`
Labels | `*.labels` | `N/A` | `key`: `value` inputs | `nil`

### Validation

We should default to the YAML view of the CR if one of the following constraints is not met:
- Every rule within every ruleGroup must either define `alert` or `record`, but not both / neither
- Every rule within every ruleGroup must define an `expr`
- Every rule within every ruleGroup that has `alert` must specify labels
- Every rule within every ruleGroup that has `alert` can additionally specify annotations
- Every rule within every ruleGroup that has `record` can additionally specify labels, but nothing else.

If any of these constraints are violated on saving an alert via the UI, we should also throw a warning.

Parent Ticket: https://github.com/rancher/dashboard/issues/1024

**Comment by lvuch** [comment]
https://xd.adobe.com/view/47515e58-7963-4f7f-927e-41cce1775fba-1202/

**Comment by jiaqiluo** [comment]
Reopen on rancher:`v2.5-0465f049780def811e00ab32067f95734f6a88f5-head` single-install with the latest UI. 

 
#### issue 1
A group should be able to have both `alertingRules` and `recordingRules` at the same time, but the UI does not allow me to have both in the same group: choosing one then another button is unclickable. 
<img width="642" alt="Screen Shot 2020-10-29 at 3 10 56 PM" src="https://user-images.githubusercontent.com/6218999/97638043-ffda1c00-19f8-11eb-826a-a212ce7a9951.png">

#### issue 2
The default value of `alert` -> `Severtiy` should be `none`
<img width="642" alt="Screen Shot 2020-10-29 at 3 11 55 PM" src="https://user-images.githubusercontent.com/6218999/97638111-20a27180-19f9-11eb-8966-adda807083c2.png">

#### issue 3
The `+` and `-` buttons should NOT show on the viewing page of an existing rule.
![image](https://user-images.githubusercontent.com/6218999/97638415-b938f180-19f9-11eb-905c-c880c5baecd5.png)

**Comment by westlywright** [comment]
@jiaqiluo there is a specific requirement in the validation of this ticket that the ui will not support both Alert rules and Alert records on one rule but not both, therefore I dont think this should be an issue. I will address the others.

**Comment by jiaqiluo** [comment]
on rancher:v2.5-94397a0c5f79394e0e0f360e011593110934ddec-head with the latest UI

- issue 1: this is expected behavior so no fixed is provided.
- issue 2: this is fixed and now the default value is `none` 
![image](https://user-images.githubusercontent.com/6218999/97921121-bc90ed80-1d17-11eb-9e85-71f7d2ad9331.png)
- issue 3: this is fixed and now there is no `+` or `-` when viewing an existing rule as form

### Reopen because of issue 4

A `string` is expected, and a `s` should be appended to the value.
For example `30s` instead of `30`

**Comment by jiaqiluo** [comment]
a new bug is found 

### issue 5: `partialResponseStrategy` should be `partial_response_strategy` 
according to https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#rulegroup

the name of the parameter should be `partial_response_strategy` in the YAML file 

<img width="984" alt="Screen Shot 2020-11-02 at 2 49 48 PM" src="https://user-images.githubusercontent.com/6218999/97923194-d7189600-1d1a-11eb-9784-faa7af800d6f.png">
<img width="469" alt="Screen Shot 2020-11-02 at 2 49 57 PM" src="https://user-images.githubusercontent.com/6218999/97923200-d97af000-1d1a-11eb-9e79-e90ed9632548.png">

**Comment by westlywright** [comment]
RE: [issue 5](https://github.com/rancher/dashboard/issues/811#issuecomment-720745043) I will be dropping this field as it was in an old design and later dropped. The field should have never been included in the first place. This should allow the user to change the value in the yaml if they wish without any naming issues.

**Comment by jiaqiluo** [comment]
The fix is validated on rancher:`v2.5-100198aff107b9a0feccdcc3e6fb5b57e83f1fd7-head` with the latest UI

- issue 4: this is fixed, and now the value of `interval` is  a string of `<value>s`
![image](https://user-images.githubusercontent.com/6218999/98019105-9a05df80-1dbe-11eb-88e9-29b6f562a025.png)

- issue 5: this is fixed, the field `partialResponseStrategy` is dropped from the UI
![image](https://user-images.githubusercontent.com/6218999/98019230-b6a21780-1dbe-11eb-8eab-82171d7876b5.png)
